### PR TITLE
Add a configurable seccomp filter on Linux

### LINKED ISSUE #5213: Add syscall filtering via seccomp for security
The Beat process on Linux can drop privileges to perform specific syscalls (like execve, fork, clone) using a seccomp-bpf filter. This is one-way operation where the process drops privileges and cannot restore them. This limits the damage that can be done if there ever would be a vulnerability.

There is a kernel version requirement, but we can test if the feature is available use it when possible.

I wrote a POC that uses google/kafel, a C library that generates the BPF syscall filter from a policy string. We could statically link with the library (we already use cgo in our Beats since 6.0).

- [Proof-of-concept GIST](https://gist.github.com/andrewkroh/1c8db5a2321ba37a728c4a453f156b48)
- [google/kafel](https://github.com/google/kafel)

When there is a policy violation the kernel kills the thread rather than the whole process. When it does this an audit event is generated.

FYI you can use `/proc/[pid]/status` to see if the process has a seccomp filter applied. Look for the `Seccomp:` line and a value > 0.

We can audit the syscalls used by a Beat by using `strace -c [beat cmd]` then develop a policy based on that data.

### PR DESCRIPTION
**Pull request by andrewkroh**
This adds a default seccomp (Secure Computing) filter on Linux that prohibits execve, execveat, fork, and vfork system calls for the Beat process. The seccomp filter policy is configurable as well. You can create your own whitelist or blacklist policy with specific actions.

Whitelist Example

```
seccomp:
  default_action: errno
  syscalls:
  - action: allow
    names:
    - open
    - close
    - read
    - exit
```

Blacklist Example

```
seccomp:
  default_action: allow
  syscalls:
  - action: log
    names:
    - execve
    - execveat
    - fork
    - vfork
  - action: kill_process     # Requires kernel 4.14+.
    names:
    - connect
    - accept
    - sendto
    - recvfrom
    - sendmsg
    - recvmsg
    - bind
    - listen
```

There is an example application in github.com/elastic/beats/libbeat/common/seccomp/sandbox for testing policies.

Closes #5213

------------------

~This is in-progress~ because I need to

- [X] write documentation
- [X] update reference configurations

Open Questions

- [x] Should I move the seccomp library into its own repository? `elastic/go-seccomp-bpf` would make sense to me. The code would be useful in other Go projects. __Update:__ _I decided to move this to it's own repo._

Other Decisions I Made

- The default policy is a blacklist because it was the most simple to start with. Whitelisting is recommended because it is more robust. `man seccomp` states "A blacklist will have to be updated whenever a potentially dangerous system call is added (or a dangerous flag or option if those are blacklisted), and it is often possible to alter the representation of a value without altering its meaning, leading to a blacklist bypass."
- I chose to not implement argument filtering. It would be useful to have for more granular policies, but I'd rather wait to see how the policies are used and see if argument filtering is required.
- The code only makes use of the `seccomp` syscall to install the filter. It does not fallback to the `prctl(2)` `PR_SET_SECCOMP` operation because it does not support flags. The `SECCOMP_FILTER_FLAG_TSYNC` flag is necessary in Go because we have little control over when threads are started and there are no guarantees that the filter would be installed before other threads start.

### CODE REVIEW

**Comment by ph** [comment]
I've did a quick check on Hound, I know they uses golint for the linter, golint seems to support to ignore out of the box generated file by looking at the code. https://github.com/golang/lint/blob/a5f4a247366d8fc436941822e14fc3eac7727015/lint.go#L122-L133

I think we don't have the [exact header](https://golang.org/s/generatedcode) to skip the file.

**Comment by andrewkroh** [comment]
@dedemorton This PR adds documentation, and I would like to get your opinion on where it should be linked into the existing content. It affects all Beats except Winlogbeat.

**Comment by dedemorton** [comment]
@andrewkroh I can understand why you are asking. Normally we would dump a topic like this into the section about configuring the Beat. Even though the seccomp settings are defined in the config file, they're kind of different from the other settings. They don't configure how the Beat behaves; they restrict what the Beat process is able to do. 

Maybe we should provide the seccomp content under the Securing xyz container (for example, [Securing Filebeat](https://www.elastic.co/guide/en/beats/filebeat/current/securing-filebeat.html). We'd need to make the introductory text a bit more generic to encompass other security-related topics (not just securing communication between Beats and other products). Would that work?
Where do you think people would expect to see the content?

In general, I'm worried that the config section has become kind of bloated, making it harder for users to find what they're looking for (I guess this is a discussion for another day).

**Comment by andrewkroh** [comment]
> Maybe we should provide the seccomp content under the Securing xyz container (for example, Securing Filebeat. We'd need to make the introductory text a bit more generic to encompass other security-related topics (not just securing communication between Beats and other products). Would that work?

@dedemorton That's exactly what I was thinking on all points. I'll start on that and have you review it.

**Comment by andrewkroh** [comment]
@dedemorton The updated documentation is ready for review. a23def6

File: libbeat/docs/security/linux-seccomp.asciidoc [hunk 1]
```diff
@@ -0,0 +1,83 @@
+[[linux-seccomp]]
+== Using Linux Secure Computing Mode (seccomp)
+
+beta[]
+
+On Linux 3.17 and newer {beatname_uc} can take advantage of secure computing
+mode, also known as seccomp, for limiting the number of system calls that the
+process can access. Specifically {beatname_uc} can install a seccomp BPF filter
+at process start-up that drops the privileges to invoke specific system calls.
```
[the remark this answers is not in the document]
Comment (andrewkroh) [reply libbeat/docs/security/linux-seccomp.asciidoc]: The Beat's main section of config doesn't support reloading so it's not possible to change it after starting the Beat.

And also seccomp BPF does not allow a filter to be removed. So I added "Once a filter is loaded by the process it cannot be removed."

File: libbeat/docs/security/linux-seccomp.asciidoc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+- `trace` - The kernel will notify a `ptrace` tracer. If no tracer is present
+  then the system call fails with `ENOSYS`  (function not implemented).
+- `trap` - The kernel will send a `SIGSYS` signal to the calling thread and not
+  execute the system call.
+- `kill_thread` - The kernel will immediately terminate the thread. Other
+  threads will continue to execute.
+- `kill_process` - The kernel will terminate the process. Available since
+  Linux 4.14.
+- `log` - The kernel will log the system call before executing it. Since Linux
+4.14. (This does not go to the Beat's log.)
```
[the remark this answers is not in the document]
Comment (andrewkroh) [reply libbeat/docs/security/linux-seccomp.asciidoc]: It depends. ™️  

I'll come back to this section after I check a few different operating systems distros to see where these logs get routed to.

**Comment by andrewkroh** [comment]
@dede I updated the documentation and reference config. Please have another look.

**Comment by adriansr** [comment]
Great job @andrewkroh

**Comment by andrewkroh** [comment]
Please use "Rebase and Merge". I have manually squashed my changes to two commits.

**Pull request merged**